Add composable views matching mode (#5173)
open-telemetry/opentelemetry-specification: specification/metrics/sdk.md

`Meter` instances have a reference to their `MeterProvider` and access
configuration only via this reference.

#### View matching mode

**Status**: [Development](../document-status.md)

A `MeterProvider` MAY accept a `view_matching_mode` parameter which controls how multiple matching [Views](#view) are applied to an [Instrument](./api.md#instrument).

Acceptable values are:

* `independent`: (Default) Each matching View creates a separate metric stream independently of any other Views registered for the same matching Instrument.

> * `composable`: Matching Views are combined (merged) to modify metric streams. See [Measurement processing](#measurement-processing) for details.

Reviewer: “Acceptable values are” does not define an SDK requirement. It leaves open whether these values must be accepted, whether extensions are allowed, and what happens for an unsupported value.

I think the parameter contract should be explicit, including the existing initialization error-handling options.

```suggestion
A `MeterProvider` MAY accept a `view_matching_mode` parameter which controls
how multiple matching [Views](#view) are applied to an
[Instrument](./api.md#instrument). A `MeterProvider` that accepts this parameter
MUST support the following values:

* `independent`: Each matching View creates a separate metric stream
  independently of any other Views registered for the same matching Instrument.
* `composable`: Matching Views are combined (merged) to modify metric streams.
  See [Measurement processing](#measurement-processing) for details.

If `view_matching_mode` is not specified, the SDK MUST use `independent`.

The SDK MAY support additional values. If an unsupported value is provided, the
SDK MUST either fail fast during initialization in accordance with the [error
handling
principles](../error-handling.md#basic-error-handling-principles), or emit a
warning, ignore the value, and use `independent`.
```

Author: > The SDK MAY support additional values.

I'm not sure about this. I would prefer the set of supported values be explicitly specified (even if some are optional for SDKs to implement).

Author: ```suggestion
A `MeterProvider` MAY accept a `view_matching_mode` parameter which controls
how multiple matching [Views](#view) are applied to an
[Instrument](./api.md#instrument). A `MeterProvider` that accepts this parameter
MUST support the following values:

* `independent`: Each matching View creates a separate metric stream
  independently of any other Views registered for the same matching Instrument.
* `composable`: Matching Views are combined (merged) to modify metric streams.
  See [Measurement processing](#measurement-processing) for details.

If `view_matching_mode` is not specified, the SDK MUST use `independent`.

If an unsupported value is provided, the SDK MUST either fail fast during
initialization in accordance with the [error handling
principles](../error-handling.md#basic-error-handling-principles), or emit a
warning, ignore the value, and use `independent`.
```

Author: Applied without "The SDK MAY support additional values.".  Feel free to reopen if you feel strongly about including that.

Reviewer: SGTM 👍

`Meter` instances have a reference to their `MeterProvider` and access
configuration only via this reference.

#### View matching mode

**Status**: [Development](../document-status.md)

A `MeterProvider` MAY accept a `view_matching_mode` parameter which controls
how multiple matching [Views](#view) are applied to an
[Instrument](./api.md#instrument). A `MeterProvider` that accepts this parameter
MUST support the following values:

* `independent`: Each matching View creates a separate metric stream
  independently of any other Views registered for the same matching Instrument.
* `composable`: Matching Views are combined (merged) to modify metric streams.
  See [Measurement processing](#measurement-processing) for details.

If `view_matching_mode` is not specified, the SDK MUST use `independent`.

> If an unsupported value is provided, the SDK MUST either fail fast during

Reviewer: To avoid the under specification of what "unsupported" means.

```suggestion
Any other value MUST be treated as unsupported.
If an unsupported value is provided, the SDK MUST either fail fast during
```

`Meter` instances have a reference to their `MeterProvider` and access
configuration only via this reference.

#### View matching mode

**Status**: [Development](../document-status.md)

A `MeterProvider` MAY accept a `view_matching_mode` parameter which controls
how multiple matching [Views](#view) are applied to an
[Instrument](./api.md#instrument). A `MeterProvider` that accepts this parameter
MUST support the following values:

* `independent`: Each matching View creates a separate metric stream
  independently of any other Views registered for the same matching Instrument.
* `composable`: Matching Views are combined (merged) to modify metric streams.
  See [Measurement processing](#measurement-processing) for details.

If `view_matching_mode` is not specified, the SDK MUST use `independent`.

> Any other value MUST be treated as unsupported.

Author: ```suggestion
```

      one View setting `aggregation` and another View setting `attribute_keys`,
      both leaving the stream `name` as the default configured by the
      Instrument). If applying the View results in conflicting metric identities

>       the implementation SHOULD apply the View and emit a warning.

Reviewer: The existing `independent` error behavior belongs in this branch. I think the whole-View fallback and advisory precedence should stay here rather than being merged into shared post-processing.

```suggestion
      the implementation SHOULD apply the View and emit a warning.
    * If applying the View would produce semantic errors (for example,
      configuring an asynchronous instrument to use the [Explicit bucket
      histogram aggregation](#explicit-bucket-histogram-aggregation)), the
      implementation SHOULD emit a warning and proceed as if that View did not
      exist.
    * If both the View and [Instrument advisory
      parameters](#instrument-advisory-parameters) specify the same aspect of
      the [Stream configuration](#stream-configuration), the setting defined by
      the View MUST take precedence over the advisory parameters.
```

Author: Why? Don't error handling and precedence ordering apply regardless of the view_matching_mode?

Edit: I see.  The common text isn't actually common.  I'll accept this.

  * (**Development**) If `view_matching_mode` is `composable`, and the Instrument could match the instrument selection criteria of one or more Views:
    * Group the matching Views by their configured stream `name`.
      * If a matching View does not configure a stream `name`, leaving `name` as null or unspecified, it applies to all streams created for that matching Instrument.

>       * Each distinct configured stream `name` creates a separate stream. If no matching View configures a stream `name`, a single stream is created using the Instrument's default name.

Reviewer: This still leaves the contents of each stream-name group implicit. A named View from another group must not affect this stream, while unnamed Views apply to every group. I think this should be explicit.

```suggestion
    * Group the matching Views by their configured stream `name`.
      * A group contains Views that configure the same stream `name` and all
        matching Views that leave `name` null or unspecified.
      * Each distinct configured stream `name` creates a separate stream. If no
        matching View configures a stream `name`, create a single group and
        stream using the Instrument's default name.
```

>     * For each resulting stream, apply the matching Views in order. For scalar aspects of the [Stream configuration](#stream-configuration) (`name`, `description`, `aggregation`, `exemplar_reservoir`, `aggregation_cardinality_limit`), the first matching View in the list that specifies that aspect wins using first-wins precedence.

Reviewer 2: Aggregation isn't stable scalar since explicit bucket histogram or base2 exponential histogram aggregations have configurable properties:

- [Base2ExponentialHistogramAggregation](https://opentelemetry.io/docs/specs/otel-config/types/#type-base2exponentialbuckethistogramaggregation)
- [ExplicitBucketHistogramAggregation](https://opentelemetry.io/docs/specs/otel-config/types/#type-explicitbuckethistogramaggregation)

I think "first matching" still applies here. I don't think we should try to merge configuration properties within Base2ExponentialHistogramAggregation or ExplicitBucketHistogramAggregation.

Author: Updated to "For all aspects of the [Stream configuration](#stream-configuration) other than `attribute_keys` (`name`, `description`, `aggregation`, `exemplar_reservoir`, `aggregation_cardinality_limit`), the first matching View in the list that specifies that aspect wins using first-wins precedence. Note that complex configurations like `aggregation` (including any internal properties such as bucket boundaries or max buckets) are treated as a single atomic unit and are not merged across Views."

>     * For each resulting stream, apply the matching Views in order. For all aspects of the [Stream configuration](#stream-configuration) other than `attribute_keys` (`name`, `description`, `aggregation`, `exemplar_reservoir`, `aggregation_cardinality_limit`), the first matching View in the list that specifies that aspect wins using first-wins precedence. Note that complex configurations like `aggregation` (including any internal properties such as bucket boundaries or max buckets) are treated as a single atomic unit and are not merged across Views.

Reviewer: `first-wins` makes View order observable, but “the list” does not define that order. I think the specification should name registration order directly.

```suggestion
    * For each resulting stream, apply the matching Views in registration order.
      For all aspects of the [Stream configuration](#stream-configuration) other
      than `attribute_keys` (`name`, `description`, `aggregation`,
      `exemplar_reservoir`, `aggregation_cardinality_limit`), the first matching
      View that specifies that aspect wins. Complex configurations like
      `aggregation` (including internal properties such as bucket boundaries or
      max buckets) are treated as a single atomic unit and are not merged across
      Views.
```

    * For `attribute_keys`, matching Views are merged by combining their attribute filters using logical AND: an attribute key is preserved if it is included in all specified allow-lists (if any) and not excluded by any specified exclude-lists.
  * For each resulting stream configured by matching View(s):
    * If applying the resolved stream configuration results in semantic errors (e.g. setting an asynchronous instrument to use the [Explicit bucket histogram aggregation](#explicit-bucket-histogram-aggregation)), the implementation SHOULD emit a warning and ignore the invalid setting. When `view_matching_mode` is `independent`, the implementation falls back to the Instrument default. When `view_matching_mode` is `composable`, the implementation falls back to the next valid setting from subsequent matching Views (or the Instrument default if no subsequent Views specify a valid setting).

>     * If both matching View(s) and [Instrument advisory parameters](#instrument-advisory-parameters) specify the same aspect of the [Stream configuration](#stream-configuration), the setting defined by the View(s) MUST take precedence over the advisory parameters.

Reviewer: This is structurally part of `composable`, but it is currently a sibling of both mode branches and mixes `independent` back in. I think this should remain nested under `composable` and only define per-setting fallback for that mode.

```suggestion
    * For each resulting stream:
      * If a setting selected for the stream would produce semantic errors, the
        implementation SHOULD emit a warning and use the next valid setting for
        that aspect from a subsequent View in the same group, or the Instrument
        default if no subsequent View specifies a valid setting.
      * If both the matching Views and [Instrument advisory
        parameters](#instrument-advisory-parameters) specify the same aspect of
        the [Stream configuration](#stream-configuration), the setting defined
        by the Views MUST take precedence over the advisory parameters.
```

      the View MUST take precedence over the advisory parameters.
  * (**Development**) If `view_matching_mode` is `composable`, and the Instrument could match the instrument selection criteria of one or more Views:
    * Group the matching Views by their configured stream `name`.
      * A group contains Views that configure the same stream `name` and all
        matching Views that leave `name` null or unspecified.
      * Each distinct configured stream `name` creates a separate stream. If no
        matching View configures a stream `name`, create a single group and
        stream using the Instrument's default name.
    * For each resulting stream, apply the matching Views in registration order.
      For all aspects of the [Stream configuration](#stream-configuration) other
      than `attribute_keys` (`name`, `description`, `aggregation`,
      `exemplar_reservoir`, `aggregation_cardinality_limit`), the first matching
      View that specifies that aspect wins. Complex configurations like
      `aggregation` (including internal properties such as bucket boundaries or
      max buckets) are treated as a single atomic unit and are not merged across
      Views.

>     * For `attribute_keys`, matching Views are merged by combining their attribute filters using logical AND: an attribute key is preserved if it is included in all specified allow-lists (if any) and not excluded by any specified exclude-lists.

Reviewer: The group contents are now defined, but the following resolution and `attribute_keys` bullets refer to “matching Views” again instead of the Views in the stream’s group.

```suggestion
    * For each resulting stream, apply the Views in its group in registration order.
      For all aspects of the [Stream configuration](#stream-configuration) other
      than `attribute_keys` (`name`, `description`, `aggregation`,
      `exemplar_reservoir`, `aggregation_cardinality_limit`), the first matching
      View that specifies that aspect wins. Complex configurations like
      `aggregation` (including internal properties such as bucket boundaries or
      max buckets) are treated as a single atomic unit and are not merged across
      Views.
    * For `attribute_keys`, merge the Views in the group by combining their attribute filters using logical AND: an attribute key is preserved if it is included in all specified allow-lists (if any) and not excluded by any specified exclude-lists.
```
